proposals/0370-modifiers.rst

```diff
@@ -122,39 +126,136 @@ Proposed Change Specification
+compatible behavior:
+
+* Only ``Multiplicity`` modifiers are permitted, and only in the places they're
+  recognized. Any use of a modifier is an error.
+
+* The kind of a modifier is determined by checking for ``Multiplicity``, not
+  through synthesis. So ``Int %m -> Bool`` is forbidden with ``-XLinearTypes
+  -XModifiers``, because ``m`` has unknown kind. But it's permitted with
+  ``-XLinearTypes -XNoModifiers``, equivalently to ``Int %(m :: Multiplicity) ->
+  Bool``.
+
+This behavior may be deprecated in future.
+
+Renaming and typechecking
+-------------------------
+Initially, the only recognized modifiers will be the multiplicity modifiers of
+linear types. These take effect during typechecking. Other modifiers are
+expected to take effect during renaming, such as the ``NoFieldSelectors``
+modifier of `proposal 512`_.
+
+.. _`proposal 512`: https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0512-nofieldselectors-per-datatype.md
+
+This poses a problem. The following is accepted::
+
+  type family F a where
+    F 1 = Many
+    F 2 = NoFieldSelectors
+
+  f :: () %(F 1) -> ()
+
+Ideally, the following would be accepted as well::
+
+  %(F 2) data G = G { g :: () }
+
+But type family resolution happens in the typechecker.
+
+Define an **RN modifier** as one that takes effect during renaming. This is
+context-dependent: ``%NoFieldSelectors`` will be an RN modifier before a
+``data`` declaration, but not before an arrow.
+
+One possible solution is to simply forbid such things for RN modifiers. If the
+type can't be resolved during renaming, it has no effect during renaming. During
+typechecking, if it gets resolved to something that would have had effect during
+renaming, we throw an error.
+
+This isn't ideal because it's not obvious to users which modifiers are RN
+modifiers.
+
+A more complicated solution involves invoking the typechecker during renaming.
+To do this we need a "stage restriction": a modifier obeys the stage restriction
+if every identifier and every type family instance in the modifier is imported,
+not defined in the current module.
+
+When renaming a modifier, if it violates the stage restriction, it takes no
+effect during renaming. We then typecheck the modifier, during typechecking. If
+it resolves to an RN modifier, we throw an error.
+
+If it doesn't violate the stage restriction, then we can typecheck it during
+renaming. If it resolves to an RN modifier, it takes effect. Subsequently, we
+typecheck it again during typechecking. We throw an error if either
+
+* It resolved to an RN modifier during renaming, and resolves to a different
+  modifier during typechecking.
+
+* It didn't resolve to an RN modifier during renaming, but does during
+  typechecking.
+
+It might resolve to a different modifier if there are overlapping instances
+defined in this module.
+
+This solution would probably not be implemented in the initial release of
```

Comment:
I think I would take this little section and move it to Alternatives. That is, let's just say, for now, that RN modifiers need not to involve type synonyms or type families. This little algorithm might allow us to improve things in the future, but I think we should wait until there is enough real pain around the status quo before doing so.